### PR TITLE
add method to IAuthenticationFactory, add interface IAzureSessionComponentListener, fix namespace for Base64UrlHelper

### PR DESCRIPTION
**Pull request by VeryEarly**

### CODE REVIEW

File: src/Common/Utilities/Base64UrlHelper.cs [hunk 1]
```diff
@@ -16,7 +16,7 @@
 using System.Globalization;
 using System.Text;
 
-namespace Microsoft.Azure.Commands.Profile.Utilities
+namespace Microsoft.WindowsAzure.Commands.Utilities.Common
```
Comment (dingmeng-xue) [inline src/Common/Utilities/Base64UrlHelper.cs:19]: Shall we change the name space?

Comment (VeryEarly) [reply src/Common/Utilities/Base64UrlHelper.cs:19]: yes, it was incorrect when it was created

File: src/Authentication.Abstractions/Models/AzureSessionEventArgs.cs [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+// ----------------------------------------------------------------------------------
+
+namespace Microsoft.Azure.Commands.Common.Authentication.Abstractions.Models
+{
+    public class AzureSessionEventArgs : EventArgs
+    {
+        public AzureSessionEventType Type;
```
Comment (erich-wang) [inline src/Authentication.Abstractions/Models/AzureSessionEventArgs.cs:20]: How about use public get and private set?

File: src/Authentication.Abstractions/AzureSession.cs [hunk 3]
```diff
@@ -242,16 +249,37 @@ public static void Modify(Action<IAzureSession> modifier)
...
                 });
         }
 
         public void ClearComponents()
         {
+            ChangeRegistry(clearHandlers);
             ChangeRegistry(_componentRegistry.Clear);
         }
 
+        private void clearHandlers()
```
Comment (erich-wang) [inline src/Authentication.Abstractions/AzureSession.cs:268]: Please use capital

File: src/Authentication.Abstractions/AzureSession.cs [hunk 4]
```diff
@@ -229,7 +231,12 @@ public static void Modify(Action<IAzureSession> modifier)
                     var key = new ComponentKey(componentName, typeof(T));
                     if (!_componentRegistry.ContainsKey(key) || overwrite)
                     {
-                        _componentRegistry[key] = componentInitializer();
+                        var component = componentInitializer();
```
Comment (erich-wang) [inline src/Authentication.Abstractions/AzureSession.cs:234]: We should remove listener in the case _componentRegistry.ContainsKey(key) && overwrite

File: src/Authentication.Abstractions/AzureSession.cs [hunk 5]
```diff
@@ -242,16 +249,37 @@ public static void Modify(Action<IAzureSession> modifier)
...
         }
 
+        private void clearHandlers()
+        {
+            _onClearContextHandler = null;
+        }
+
+        public void ClearComponentsOnClearAzContext()
+        {
+            this.OnClearAzContext(new AzureSessionEventArgs(AzureSessionEventType.OnClearContext));
```
Comment (erich-wang) [inline src/Authentication.Abstractions/AzureSession.cs:275]: nit: suggest to remove `this` to keep code style consistent

File: src/Authentication.Abstractions/Interfaces/IAzureSession.cs [hunk 6]
```diff
@@ -126,6 +126,11 @@ public interface IAzureSession : IExtensibleModel
         /// </summary>
         void ClearComponents();
 
+        /// <summary>
+        /// Remove all components from the session shared component registry with on clear-azcontext event
+        /// </summary>
+        void ClearComponentsOnClearAzContext();
```
Comment (erich-wang) [inline src/Authentication.Abstractions/Interfaces/IAzureSession.cs:132]: better name?

File: src/Authentication.Abstractions/Models/AzureSessionEventType.cs [hunk 7]
```diff
@@ -0,0 +1,21 @@
...
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+// ----------------------------------------------------------------------------------
+
+namespace Microsoft.Azure.Commands.Common.Authentication.Abstractions.Models
+{
+    public enum AzureSessionEventType
+    {
+        OnClearContext
```
Comment (erich-wang) [inline src/Authentication.Abstractions/Models/AzureSessionEventType.cs:19]: How about use ContextCleared? Onxxx is more like function name

**Review by erich-wang — CHANGES_REQUESTED** [review]

File: src/Authentication.Abstractions/AzureSession.cs [hunk 8]
```diff
@@ -242,16 +249,37 @@ public static void Modify(Action<IAzureSession> modifier)
...
+        }
+
+        protected virtual void OnClearAzContext(AzureSessionEventArgs e)
+        {
+            _onClearContextHandler?.Invoke(this, e);
```
Comment (isra-fel) [inline src/Authentication.Abstractions/AzureSession.cs:280]: We can call `handler.Invoke()` directly maybe? No need to wrap it in a method

**Review by isra-fel — CHANGES_REQUESTED** [review]

File: src/Authentication.Abstractions/Interfaces/IAzureSession.cs (hunk 6 above)
Comment (VeryEarly) [reply src/Authentication.Abstractions/Interfaces/IAzureSession.cs:132]: RaiseClearAzContext

File: src/Authentication.Abstractions/Models/AzureSessionEventType.cs (hunk 7 above)
Comment (VeryEarly) [reply src/Authentication.Abstractions/Models/AzureSessionEventType.cs:19]: ContextCleared

File: src/Authentication.Abstractions/Interfaces/IAzureSession.cs [hunk 9]
```diff
@@ -126,6 +126,11 @@ public interface IAzureSession : IExtensibleModel
         /// </summary>
         void ClearComponents();
 
+        /// <summary>
+        /// Remove all components from the session shared component registry with on clear-azcontext event
```
Comment (VeryEarly) [inline src/Authentication.Abstractions/Interfaces/IAzureSession.cs:130]: update comments

File: src/Authentication.Abstractions/AzureSession.cs (hunk 4 above)
Comment (VeryEarly) [reply src/Authentication.Abstractions/AzureSession.cs:234]: this logic was covered by || overwrite isn't it?

Comment (VeryEarly) [reply src/Authentication.Abstractions/AzureSession.cs:234]: > We should remove listener in the case _componentRegistry.ContainsKey(key) && overwrite

this logic was covered by || overwrite isn't it?

**Review by isra-fel — APPROVED** [review]

**Pull request merged**